bigtable/client/internal/common_client.h

```diff
@@ -44,25 +49,44 @@ class CommonClient {
    * and/or when the credentials require explicit refresh.
    */
   void reset() {
-    stub_.reset();
-    channel_.reset();
+    std::lock_guard<std::mutex> lk(mu_);
+    stubs_.clear();
   }
 
   StubPtr Stub() {
-    if (not stub_) {
-      auto channel = grpc::CreateCustomChannel(Traits::Endpoint(options_),
-                                               options_.credentials(),
-                                               options_.channel_arguments());
-      stub_ = Interface::NewStub(channel);
-      channel_ = channel;
+    std::unique_lock<std::mutex> lk(mu_);
+    if (stubs_.empty()) {
+      // Release the lock while making remote calls.  gRPC uses the current
+      // thread to make remote connections (and probably authenticate), holding
+      // a lock for long operations like that is a bad practice.  Releasing
+      // the lock here can result in wasted work, but that is a smaller problem
```

Comment:
The wasted work being more stubs than necessary getting created, as multiple threads attempt to initialize the pool?

This might be ok but there is also extra work in the CFE/GFE to establish these channels. So it seems possible that one application with many threads and therefore a large connection pool size could create and then tear down hundreds of gRPC connections...

Reply:
> The wasted work being more stubs than necessary getting created, as multiple threads attempt
> to initialize the pool?

Yes.

> This might be ok but there is also extra work in the CFE/GFE to establish these channels. So it
> seems possible that one application with many threads and therefore a large connection pool
> size could create and then tear down hundreds of gRPC connections...

I was going to say that was true, and changed the code, but then I remembered.  gRPC only creates one socket per endpoint+attributes combo.  That is why we have to set some artificial attributes in `CreateChannelPool()`.  I do not think this is 100% safe, but "mostly harmless".